web/src/composebox_typeahead.ts

```diff
@@ -1039,54 +1305,59 @@ export function content_typeahead_selected(item, query, input_element, event) {
     return beginning + rest;
 }
 
-export function compose_content_matcher(completing, token) {
-    switch (completing) {
+export function compose_content_matcher(item: TypeaheadSuggestion, token: string): boolean {
+    switch (item.type) {
         case "emoji":
-            return typeahead.get_emoji_matcher(token);
+            return typeahead.get_emoji_matcher(token)(item);
         case "slash":
-            return get_slash_matcher(token);
+            return get_slash_matcher(token)(item);
         case "stream":
-            return get_stream_or_user_group_matcher(token);
+            return get_stream_or_user_group_matcher(token)(item);
         case "syntax":
-            return get_language_matcher(token);
+            return get_language_matcher(token)(item);
         case "topic_list":
-            return get_topic_matcher(token);
+            return get_topic_matcher(token)(item);
+        case "topic_jump":
+        case "time_jump":
+            // these don't actually have a typeahead popover, so we return quickly here.
+            return true;
+        default:
+            return false;
     }
-
-    return function () {
-        switch (completing) {
-            case "topic_jump":
-            case "time_jump":
-                // these don't actually have a typeahead popover, so we return quickly here.
-                return true;
-            default:
-                return undefined;
-        }
-    };
 }
 
-export function sort_results(completing, matches, token) {
-    switch (completing) {
+// This is an exhaustive switch statement so we don't need to return after it.
+// eslint-disable-next-line consistent-return
```

Comment:
Another option would be

```ts
import {assertNever} from "assert-never";

switch (matches.type) {
    …
    default:
        return assertNever(matches);
}
```

which preserves compile-time exhaustiveness checking because [`assertNever`](https://www.npmjs.com/package/assert-never) takes an argument of type `never`.

Reply:
Ooo I like that, thanks!